sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala

```diff
@@ -85,8 +85,15 @@ private[sql] object FileSourceStrategy extends Strategy with Logging {
         ExpressionSet(normalizedFilters.filter(_.references.subsetOf(partitionSet)))
       logInfo(s"Pruning directories with: ${partitionKeyFilters.mkString(",")}")
 
-      val dataColumns =
-        l.resolve(fsRelation.dataSchema, fsRelation.sparkSession.sessionState.analyzer.resolver)
+      val dataColumns = l.resolve(fsRelation.dataSchema,
+        fsRelation.sparkSession.sessionState.analyzer.resolver).map { c =>
+          fsRelation.dataSchema.find(_.name == c.name).map { f =>
+            c match {
+              case a: AttributeReference => a.withMetadata(f.metadata)
+              case _ => c
+            }
+          }.getOrElse(c)
+        }
```

Comment:
When I add a regression test, I find that this change is not needed. Not sure which recent commit helps keeping metadata after resolving. I will add the regression test still.